### PR TITLE
Refactor rootston keyboard

### LINKED ISSUE #398: keyboard_keysym_press called twice
`keyboard_keysym_press` is called several times per keypress: one time in `keyboard_keysyms_simple` and one time in `keyboard_keysyms_xkb`. So pressing Alt+Shift+2 will add 2 _and_ @ in `roots_keyboard_pressed_keysyms`.

Introduced in #394.

cc @Ongy

### PR DESCRIPTION
**Pull request by emersion**
Test plan: keybinds such as `Ctrl+Shift+E` shouldn't work anymore. Keybinds such as `Ctrl+Shift+e` and `Ctrl+E` should work. Check that switching VTs works.

Fixes #398

### CODE REVIEW

File: rootston/keyboard.c [hunk 1]
```diff
@@ -56,21 +75,13 @@ static void keyboard_binding_execute(struct roots_keyboard *keyboard,
...
 
 /**
- * Process a keypress from the keyboard.
+ * Execute a built-in, hardcoded compositor action when a keysym is pressed.
  *
  * Returns true if the keysym was handled by a binding and false if the event
  * should be propagated to clients.
  */
-static bool keyboard_keysym_press(struct roots_keyboard *keyboard,
+static bool keyboard_press_keysym(struct roots_keyboard *keyboard,
```
Comment (acrisci) [inline rootston/keyboard.c]: Probably needs a better name like `keyboard_keysym_compositor_action()`.

**Comment by acrisci** [comment]
The keyboard command "close" should be seat aware. It should close the focused window for the seat of the keyboard device that executed the binding.

**Comment by acrisci** [comment]
We should probably have a focus list per seat so things like "next_window" will work correctly. Would also affect reverting focus when the focused view is closed for a seat. (these are probably separate issues though).

File: rootston/keyboard.c [hunk 2]
```diff
@@ -116,86 +155,64 @@ static bool keyboard_keysym_press(struct roots_keyboard *keyboard,
...
-	bool handled = keyboard_keysyms_xkb(keyboard, keycode, event->state);
+	xkb_keycode_t keycode = event->keycode + 8;
 
+	uint32_t modifiers;
+	const xkb_keysym_t *keysyms;
+	size_t keysyms_len = keyboard_keysyms_translated(keyboard, keycode,
+		&keysyms, &modifiers);
+	bool handled = keyboard_handle_keysyms(keyboard,
+		keyboard->pressed_keysyms_translated, modifiers, keysyms, keysyms_len,
+		event->state);
```
Comment (acrisci) [inline rootston/keyboard.c]: There is a subtle bug here with this approach. If I bind to `a` and `b` and then press `a`, press `b`, release `a`, release `b`, then the `a` binding will run twice, once when I press `a` and once when I press `b` when I would expect it to only run each binding once.

File: rootston/keyboard.c [hunk 3]
```diff
@@ -90,7 +101,35 @@ static bool keyboard_keysym_press(struct roots_keyboard *keyboard,
...
+/**
+ * Press or release keysyms.
+ *
+ * Returns true if the keysym was handled by a binding and false if the event
+ * should be propagated to clients.
+ */
+static bool keyboard_handle_keysyms(struct roots_keyboard *keyboard,
+		xkb_keysym_t *pressed_keysyms, uint32_t modifiers,
+		const xkb_keysym_t *keysyms, size_t keysyms_len,
+		enum wlr_key_state state) {
```
Comment (acrisci) [inline rootston/keyboard.c]: Could use a better name that indicates that it actually executes bindings like `keyboard_execute_bindings()`.

File: rootston/keyboard.c [hunk 4]
```diff
@@ -116,86 +155,64 @@ static bool keyboard_keysym_press(struct roots_keyboard *keyboard,
...
 /*
- * Process keypresses from the keyboard as xkb sees them.
+ * Get keysyms and modifiers from the keyboard as xkb sees them.
  *
  * This uses the xkb keysyms translation based on pressed modifiers and clears
  * the consumed modifiers from the list of modifiers passed to keybind
  * detection.
  *
- * (On US layout) this will trigger: [Alt]+[at]
+ * On US layout, this will trigger: Alt + @
```
Comment (acrisci) [inline rootston/keyboard.c]: I don't understand this comment line.

Comment (acrisci) [reply rootston/keyboard.c]: Oh I see, the context for this comment comes from several lines below. This is kind confusing to read and should probably be made self contained for each.

File: rootston/keyboard.c (hunk 2 above)
Comment (acrisci) [reply rootston/keyboard.c]: I think the easy fix to this is to make sure the keysyms pressed are *exactly* the ones specified by the binding.

File: rootston/keyboard.c [hunk 5]
```diff
@@ -116,86 +155,64 @@ static bool keyboard_keysym_press(struct roots_keyboard *keyboard,
...
-
-	bool handled = keyboard_keysyms_xkb(keyboard, keycode, event->state);
+	xkb_keycode_t keycode = event->keycode + 8;
 
+	uint32_t modifiers;
+	const xkb_keysym_t *keysyms;
+	size_t keysyms_len = keyboard_keysyms_translated(keyboard, keycode,
+		&keysyms, &modifiers);
+	bool handled = keyboard_handle_keysyms(keyboard,
+		keyboard->pressed_keysyms_translated, modifiers, keysyms, keysyms_len,
```
Comment (acrisci) [inline rootston/keyboard.c]: These variables (`keyboard->pressed_keysyms_*`) are set but never used.

Comment (acrisci) [reply rootston/keyboard.c]: Disregard that, I see how they are used now.

**Comment by acrisci** [comment]
Is there a good reason why keysym translation is a rootston responsibility? Why is this state not kept on wlr-keyboard itself?

**Comment by emersion** [comment]
>Is there a good reason why keysym translation is a rootston responsibility? Why is this state not kept on wlr-keyboard itself?

The seat takes keycodes, not keysyms. Also, this whole story about using modifiers or not when getting keysyms from keycodes shows that this compositor should be able to customize how it handles keycodes.

**Comment by emersion** [comment]
Refactored a little bit more, this should address all comments.

**Comment by acrisci** [comment]
👍 looks good.

**Pull request merged**